### ISSUE TITLE
WebAssembly version of the validator

### ISSUE THREAD

**Issue by monochromer** [validator/validator#1651]
**Is your feature request related to a problem? Please describe.**
Since the VNU validator is often used as a frontend tool, it would be convenient to create a version suitable for JavaScript environments.

**Describe the solution you'd like**
Is it possible to create a WebAssembly version of the VNU validator and a JavaScript wrapper over it?

**Comment by monochromer** [comment]
[WebAssembly 3.0](https://webassembly.org/news/2025-09-17-wasm-3.0/) with garbage collection is now available.

**Comment by sideshowbarker** [comment]
I agree it could be great to have a WebAssembly version of the validator, but there’s very little chance I’d be able to make time myself any time soon to create it.

So, to help this feature request get more visibility I’ve labeled it with **` help wanted `** and pinned it to the top of the issues page.

**Comment by sideshowbarker** [comment]
I’m going to (reluctantly) close this issue — because, after spending some time investigating it myself, my conclusion is that it’s currently not possible to convert vnu into WebAssembly, and it’s basically going to remain so into the foreseeable future.

And given that, I don’t want anybody else to end up unnecessarily wasting time investigating this — because I’m certain they’re going to end up discovering what I did.

The fundamental problem is just that vnu has a lot of dependencies, and the existing tools for converting from Java to WebAssembly all choke on those dependencies. To name some specifics:

- org.xml.sax — in particular: InputSource, SAXParseException, SAXException, XMLReader, SAXParserFactory, AttributesImpl
- ICUJ
- Jing

And long with those dependencies, there are some even-more fundamental issues — such as:

- the checker code makes use of multithreading, and apparently that’s … just not going to work
- the checker code uses java.lang.ClassLoader, and apparently that also is just not going to work

**Comment by monochromer** [comment]
Thanks for your investigating
